### PR TITLE
[WIP] Fix issues with number format

### LINKED ISSUE #33798: Price on product view page becomes 0,00 after page has finished loading
### Preconditions (*)

1. Magento 2.4.3
2. PHP 7.4
3. Country Options is set to US
4. Locale Options (Local) is set to Germany
5. Base Currency: Euro
6. Default Display Currency: Euro
7. Allowed Currencies: Euro
8. 3 different StoreViews are set up

### Steps to reproduce (*)

1. Create new product (price doesn't matter)
2. Go to product list page (categorie view page)
3. Go to product details page (product view page)

### Expected result (*)

1. Actual price on product details page should be shown

### Actual result (*)

1. Price on product list page is shown correctly
2. But on product details page, the page loads, correct price is shown for a millisecond but when the page has finished loading, price is shown as "0,00 €"

It seems like there is some JS or anything like that going on, when the page has finished loading, the price is set "0,00 €"
**Important: This bug is only present in 2.4.3. I downgraded to 2.4.2 and there, everything works fine.**

I found an old, _some kind of related_ issue which was fixed: https://github.com/magento/magento2/issues/4077
My store is set to **Locale Options (Local) is set to Germany**. _But I did not test if this is actually causing the issue. I will test as soon as possible._

### LINKED ISSUE #33856: Currency Symbol Disappear on Product View Page (After Fully Loaded)
### Preconditions (*)

1. Magento 2.4.3

### Steps to reproduce (*)

1. Visit any product page
2. Currency symbol appears eg - $1.00
3. Wait untill page fully load.
4. Currency symbol disappear eg - 1.00

### Expected result (*)

1. Currency Symbol to remain displayed.

### Actual result (*)

1. Currency symbol disappear

Tried with a freshly installed Magento 2.4.3. I can confirm this is a bug.
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._

### LINKED ISSUE #33910: [Issue] Product Price Showing ZAR 0.00 when South African Currency Used
### Preconditions (*)
    Magento 2.4.3
    PHP Version 7.4.22
    Elasticsearch 7.9
    Vanilla, clean installation with sample-data

### Steps to reproduce (*)
   
1. Go to Admin -> Store -> Configuration -> Currency Setup -> Set Base Currency to South African Rand
2. Go to Admin -> Store -> Configuration -> Currency Setup -> Set Default Currency to South African Rand
3. Reindex and Flush Cache

### Expected result (*)

1. Price of the product should show as ZAR xxx (where xxx is the price).

### Actual result (*)
1. Price of the product shows as ZAR 0.00. An inspection of the value shows that ZAR 0.00 but data-price-amount is correct.
ZAR

P.S Changing back to USD shows the product price correctly.

### LINKED ISSUE #34139: Frontend Prices is 0 After upgrading from 2.4.2 to 2.4.3
### Preconditions (*)

1. Magento Version 2.4.3
2. Upgrade My Application from Version 2.4.2 to 2.4.3 Using Composer

### Steps to reproduce (*)

1. Upgrade Magento Version from 2.4.2 to 2.4.3
2. When open and product page the price appears normal then when the page completely loaded the price transferred to 0
3. When open checkout all shipping prices and product prices appears 0

Note : I install a clean Magento Version then install my themes and modules with an empty database and create some demo products till here everything is working fine but when I change the database to the main one and run the commands to upgrade it then the error mentioned in point 2 and 3 occured.

### Expected result (*)

1. Magento with Version 2.4.2 [https://prnt.sc/1t7h967](https://prnt.sc/1t7h967)

### Actual result (*)

1. Magento with Version 2.4.3 [https://prnt.sc/1t7hfzc](https://prnt.sc/1t7hfzc)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._

### LINKED ISSUE #34229: Configurable products has zero price (in product view) after Locale is changed to Greek.
### Preconditions (*)

1. Plesk Obsidian 18.0.38
2. Magento Open Source 2.4.3
3. Php 7.4.23
4. Select English (United State) Locale from Stores --> Configuration --> General --> General (http://prntscr.com/1un7u6e)

### Steps to reproduce (*)

1. Change Catalog Input Type for Store Owner to Visual Swatch, for product attribute color (http://prntscr.com/1un5ram)
2. Create some options (http://prntscr.com/1un5v29) 
3. Add new configurable product from admin area (http://prntscr.com/1un63k5)
4. Add sample data to parent product(http://prntscr.com/1un6cec)
5. Create configurations

> - Select attribute color and click Next
> - Select all attribute values and click next
> - Skip image uploading for all SKUs
> - Apply single price to all SKUs (e.g: 50)
> - Apply single quantity to each SKUs (e.g: 120) and click Next
> - Click Generate Products

6. Save product (http://prntscr.com/1un7092)	
7. As a result the price of each product seems to be normal (http://prntscr.com/1un8fr9 , http://prntscr.com/1un8i11)
8. Change store locale from English (United State) to Greek (Greece) : http://prntscr.com/1un8o7r
9. Flush Magento Cache

### Expected result (*)

The product prices should be the same with the result of English Locale (http://prntscr.com/1un8fr9 , http://prntscr.com/1un8i11)

### Actual result (*)

All prices are zero (http://prntscr.com/1un9h1e , http://prntscr.com/1un9ii1)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by ihor-sviziev**
### Description (*)
This Pull Request fixes the listed below issues.

The reason why they appear - in the different ICU library versions.

The script that I used for testing:
```bash
cat - <<'EOF' | php
<?php

try {
    $reflector = new \ReflectionExtension('intl');
    ob_start();
    $reflector->info();
    $output = strip_tags(ob_get_clean());
    preg_match('/^ICU version (?:=>)?(.*)$/m', $output, $matches);
    $icuVersion = trim($matches[1]);
} catch (\ReflectionException $e) {
    $icuVersion = null;
}
echo 'ICU Version: ' . $icuVersion . PHP_EOL;


$formatter = new NumberFormatter('de_DE', NumberFormatter::CURRENCY);
echo 'Original pattern: ' . $formatter->getPattern() . PHP_EOL;
echo 'Original result: ' . $formatter->formatCurrency(10.3, "EUR") . PHP_EOL;

$formatter->setSymbol(\NumberFormatter::CURRENCY_SYMBOL, '');
echo 'Pattern with replaced currency symbol: ' . $formatter->getPattern() . PHP_EOL;
echo 'Result with changed currency symbol: ' . $formatter->formatCurrency(10.3, "EUR") . PHP_EOL;

$formatter->setPattern( str_replace('¤','', $formatter->getPattern() ) );
echo 'Pattern with pattern: ' . $formatter->getPattern() . PHP_EOL;
echo 'Result with the changed pattern: ' . $formatter->formatCurrency(10.3, "EUR") . PHP_EOL;

EOF
```

Originally, Magento used replacing the currency symbol.
My PR adds removing the currency symbol from the pattern.

The output on different versions for the de_DE locale and EUR locale:
- Version 65.1:
✔ Result with changed currency symbol (original implementation): 10,30
✔ Result with the changed pattern (on this PR): 10,30
 

    ```
    ICU Version: 65.1
    Original pattern: #,##0.00 ¤
    Original result: 10,30 €
    Pattern with replaced currency symbol: #,##0.00 ¤
    Result with changed currency symbol: 10,30 
    Pattern with pattern: #,##0.00 
    Result with the changed pattern: 10,30 
    ```

- Version 50.2:
❌ Result with changed currency symbol (original implementation): 10,30 €
✔ Result with the changed pattern (on this PR): 10,30

    ```
    ICU Version: 50.2
    Original pattern: #,##0.00 ¤
    Original result: 10,30 €
    Pattern with replaced currency symbol: #,##0.00 ¤
    Result with changed currency symbol: 10,30 €
    Pattern with pattern: #,##0.00 
    Result with the changed pattern: 10,30
    ```

But, on the fa_IR locate with USD currency:
```
cat - <<'EOF' | php
<?php

try {
    $reflector = new \ReflectionExtension('intl');
    ob_start();
    $reflector->info();
    $output = strip_tags(ob_get_clean());
    preg_match('/^ICU version (?:=>)?(.*)$/m', $output, $matches);
    $icuVersion = trim($matches[1]);
} catch (\ReflectionException $e) {
    $icuVersion = null;
}
echo 'ICU Version: ' . $icuVersion . PHP_EOL;


$formatter = new NumberFormatter('fa_IR', NumberFormatter::CURRENCY);
echo 'Original pattern: ' . $formatter->getPattern() . PHP_EOL;
echo 'Original result: ' . $formatter->formatCurrency(0, "USD") . PHP_EOL;

$formatter->setSymbol(\NumberFormatter::CURRENCY_SYMBOL, '');
echo 'Pattern with replaced currency symbol: ' . $formatter->getPattern() . PHP_EOL;
echo 'Result with changed currency symbol: ' . $formatter->formatCurrency(0, "USD") . PHP_EOL;

$formatter->setPattern( str_replace('¤','', $formatter->getPattern() ) );
echo 'Pattern with pattern: ' . $formatter->getPattern() . PHP_EOL;
echo 'Result with the changed pattern: ' . $formatter->formatCurrency(0, "USD") . PHP_EOL;

EOF
```

- Version 65.1:
✔ Result with changed currency symbol (original) is correct: ۰٫۰۰ (equivalent to **0.00**)
❌ Result with the changed pattern (on this PR) is not correct: ۰ (equivalent to **0**)
    ```
    ICU Version: 65.1
    Original pattern: ‎¤#,##0
    Original result: ‎$۰٫۰۰
    Pattern with replaced currency symbol: ‎¤#,##0
    Result with changed currency symbol: ‎۰٫۰۰
    Pattern with pattern: ‎#,##0
    Result with the changed pattern: ‎۰
    ```

- Version 50.2:
❌ Result with changed currency symbol (original) is not correct: US$۰٫۰۰ (equivalent to **US$0.00**)
✔ Result with the changed pattern (on this PR) is correct: ۰٫۰۰ (equivalent to **0.00**)
    ```
    ICU Version: 50.2
    Original pattern: ‎¤#,##0;‎(¤#,##0)
    Original result: ‎US$۰٫۰۰
    Pattern with replaced currency symbol: ‎¤#,##0;‎(¤#,##0)
    Result with changed currency symbol: ‎US$۰٫۰۰
    Pattern with pattern: ‎#,##0.00;‎(#,##0.00)
    Result with the changed pattern: ‎۰٫۰۰
    ```

### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/magento2/issues/33798
2. Fixes https://github.com/magento/magento2/issues/33856
3. Fixes https://github.com/magento/magento2/issues/33910
4. Fixes https://github.com/magento/magento2/issues/34139
5. Fixes https://github.com/magento/magento2/issues/34229

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Directory/Model/Currency.php [hunk 1]
```diff
@@ -431,6 +431,7 @@ private function formatCurrency(string $price, array $options): string
         if ((array_key_exists(LocaleCurrency::CURRENCY_OPTION_DISPLAY, $options)
                 && $options[LocaleCurrency::CURRENCY_OPTION_DISPLAY] === \Magento\Framework\Currency::NO_SYMBOL)) {
             $formattedCurrency = str_replace(' ', '', $formattedCurrency);
+            $formattedCurrency = str_replace(' ', '', $formattedCurrency);
```
Comment (Bashev) [inline app/code/Magento/Directory/Model/Currency.php]: What is the idea here with duplicated line?

Comment (ihor-sviziev) [reply app/code/Magento/Directory/Model/Currency.php]: Oh, that is mistake

Comment (ihor-sviziev) [reply app/code/Magento/Directory/Model/Currency.php]:
```suggestion
```

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Looks like this solution doesn't work on the locale:

```
cat - <<'EOF' | php
<?php

try {
    $reflector = new \ReflectionExtension('intl');
    ob_start();
    $reflector->info();
    $output = strip_tags(ob_get_clean());
    preg_match('/^ICU version (?:=>)?(.*)$/m', $output, $matches);
    $icuVersion = trim($matches[1]);
} catch (\ReflectionException $e) {
    $icuVersion = null;
}
echo $icuVersion . PHP_EOL;


$formatter = new NumberFormatter('fa_IR', NumberFormatter::CURRENCY);
echo $formatter->getPattern() . PHP_EOL;
echo $formatter->formatCurrency(0, "EUR") . PHP_EOL;

$formatter->setSymbol(\NumberFormatter::CURRENCY_SYMBOL, '');
echo $formatter->getPattern() . PHP_EOL;
echo $formatter->formatCurrency(0, "EUR") . PHP_EOL;

$formatter->setPattern( str_replace('¤','', $formatter->getPattern() ) );
echo $formatter->getPattern() . PHP_EOL;
echo $formatter->formatCurrency(0, "EUR") . PHP_EOL;

EOF
```
This code on the 0 value has different behavior on difference ICU versions (see the latest line):
```
50.2
‎¤#,##0;‎(¤#,##0)
‎€۰٫۰۰
‎¤#,##0;‎(¤#,##0)
‎€۰٫۰۰
‎#,##0.00;‎(#,##0.00)
‎۰٫۰۰
```
and
```
65.1
‎¤#,##0
‎€۰٫۰۰
‎¤#,##0
‎۰٫۰۰
‎#,##0
‎۰
```

**Comment by ihor-sviziev** [comment]
Hi @adifucan @dhorytskyi,
Could you take a look at these issues?
I tried to fix them, but some new issues appear on different ICU library versions (not sure if that's the real reason)

**Comment by Bashev** [comment]
@ihor-sviziev 

Issue is related to NumberFormatter and method setSymbol, it [not works for formatCurrency](https://www.php.net/manual/en/numberformatter.setsymbol.php#124153)

Issue can be solved with adding code below on line 433.
`$formattedCurrency = str_replace($this->getCurrencySymbol(), '', $formattedCurrency);`

or remove lines between 368 and 370 
```
if ($this->canUseNumberFormatter($options)) {
   return $this->formatCurrency($price, $options);
}
```

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@Bashev, thx for the tip - looks like it should work.
Let's see if the tests will work

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@Bashev, it looks like it won't work correctly on the ICU version 50.2. Currently, it shows:
```
Original result: ‎US$۰٫۰۰
Result with changed currency symbol: ‎US$۰٫۰۰
Result with the changed pattern: ‎۰٫۰۰

```
While version 65.1:
```
Original result: ‎$۰٫۰۰
Result with changed currency symbol: ‎۰٫۰۰
Result with the changed pattern: ‎۰

```

Feels like we have a method that returns the currency symbol, but... feels like different ICU versions even return different symbols:
I took `en-US@currency=JPY:` locale for detecting correctly the localized currency symbol:

```
cat -<<'EOF' | php
<?php
$locale='fa_IR'; //browser or user locale
$currency='USD';
$fmt = new NumberFormatter( $locale."@currency=$currency", NumberFormatter::CURRENCY );
$symbol = $fmt->getSymbol(NumberFormatter::CURRENCY_SYMBOL);
echo $symbol . PHP_EOL;
EOF
```

On 50.2:
```
US$
```

On 65.1:
```
$
```

**Comment by Bashev** [comment]
@ihor-sviziev,
What happens with this version of ICU if you totally "ignore" \NumberFormatter with removing this from the model?

```
if ($this->canUseNumberFormatter($options)) {
   return $this->formatCurrency($price, $options);
}
```

I cannot test with this version of ICU.

**Comment by ihor-sviziev** [comment]
@Bashev,
 the issue with that, that that was performance improvement... I don't think it's a good idea to revert it...

**Comment by zlep** [comment]
Hi, any update on this? I'm still looking to start with 2.4.3 instead of 2.4.2 (which was the last version without this bug).

**Comment by ihor-sviziev** [comment]
@zlep,
Unfortunately, this solution still has some bugs. 
There is a working workaround, you can use it in the meanwhile https://github.com/magento/magento2/issues/33856#issuecomment-904606945

**Comment by zlep** [comment]
@ihor-sviziev Yes, thanks. But it's still worked on, isn't it?

**Comment by ihor-sviziev** [comment]
Unfortunately, the solution doesn't work fine on all locales and ICU
versions

**Comment by Bashev** [comment]
> 
> 
> @Bashev,
> the issue with that, that that was performance improvement... I don't think it's a good idea to revert it...

@ihor-sviziev Is it a good idea to check ICU version in "canUseNumberFormatter" method? This way will keep performance improvement for environments which have compatible ICU version, but also will have backward compatibility.

**Comment by ihor-sviziev** [comment]
@Bashev, currently don't have enough time to check if it will work in all cases, but probably it should work

**Comment by ihor-sviziev** [comment]
Closed in favor of https://github.com/magento/magento2/pull/34156

**Comment by ihor-sviziev** [comment]
Re-opened because of https://github.com/magento/magento2/pull/34156#issuecomment-930620874

**Comment by ihor-sviziev** [comment]
Looks like this issue was fixed in https://github.com/magento/magento2/commit/ae304d43cd36322b1ff8145d70c9b2e9aeb0cde
I’m closing this PR as not relevant anymore

**Pull request closed without merging**